### ISSUE TITLE
Change in exception hierarchy

### ISSUE THREAD

**Issue by reaperhulk** [kjd/idna#43]
The fix for #36 (https://github.com/kjd/idna/commit/ebefacd3134d0f5da4745878620a6a1cba86d130) changes the exception hierarchy and breaks several downstream dependencies (twisted, pyca/cryptography, possibly more). Looking at the diff, would it be possible to revert and simply catch `UnicodeEncodeError` rather than `UnicodeError`? `UnicodeEncodeError` will be raised by `encode`, which is what you're trying to catch, while you don't want to catch any of the IDNAErrors. Alternately, this ugly code (which could be improved) also passes the tests:

```python
def alabel(label):

    if not label:
        raise IDNAError('No Input')

    if not valid_label_length(label):
        raise IDNAError('Label too long')

    try:
        label = label.encode('ascii')
        encoded = True
    except UnicodeError:
        encoded = False

    if encoded is True:
        try:
            ulabel(label)
            return label
        except IDNAError:
            raise IDNAError('The label {0} is not a valid A-label'.format(label))

    label = unicode(label)
    check_label(label)
    label = _punycode(label)
    label = _alabel_prefix + label

    if not valid_label_length(label):
        raise IDNAError('Label too long')

    return label

```

**Comment by kjd** [comment]
Since you are not the first to note this unintended consequence (https://github.com/kjd/idna/commit/ebefacd3134d0f5da4745878620a6a1cba86d130) I'll try and get a new release out the door in the next day that makes IDNAError a descendant of UnicodeError again. I'm surprised so many dependencies are looking for UnicodeError when that was never documented.

**Comment by kjd** [comment]
Fixed in v2.4, now released.

**Comment by reaperhulk** [comment]
Thanks, I can confirm this resolves the issue for cryptography!
